src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,258 @@
+        """
+        """
+        paths = self.process_samples(itr, paths)
+        self.episode_rewards.extend([
+            path for path, complete in zip(paths['undiscounted_returns'],
+                                           paths['complete']) if complete
+        ])
+        self.success_history.extend([
+            path for path, complete in zip(paths['success_history'],
+                                           paths['complete']) if complete
+        ])
+        last_average_return = np.mean(self.episode_rewards)
+        if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:  # noqa: E501
+            for gradient_step in range(self.gradient_steps):
+                    samples = self.replay_buffer.sample(self.buffer_batch_size)
+                    # self.update_q_functions(itr, samples)
+                    self.optimize_policy(gradient_step, samples)
+                    self.update_targets()
+            tabular.record('reward', last_average_return)
+
+        return last_average_return
+
+    def optimize_policy(self, itr, samples):
+        """ Optimize the policy based on the policy objective from the sac paper.
+
+        Args:
+            itr (int) - current training iteration
+            samples() - samples recovered from the replay buffer
+        Returns:
+            None
+        """
+
+        obs = samples["observation"]
+        actions = samples["action"]
+        rewards = samples["reward"]
+        terminals = samples["terminal"]
+        next_obs = samples["next_observation"]
+
+        action_dists = self.policy(torch.Tensor(obs))
+        if ((itr % (self.gradient_steps - 1) == 0) and itr != 0):
+            entropy = action_dists.entropy()
+            tabular.record("policy/mu_entropy", torch.mean(entropy).detach().numpy().item())
+            tabular.record("policy/mu_stdev", torch.mean(action_dists.variance.detach().flatten()).item()**0.5)
+            tabular.record("policy/mu_mean", torch.mean(action_dists.mean.detach()).item())
+        new_actions = action_dists.rsample()
+        log_pi = action_dists.log_prob(new_actions)
+
+        # tune the coefficient alpha
+        if self.use_automatic_entropy_tuning:
+            alpha_loss = (-(self.log_alpha)* (log_pi.detach() + self.target_entropy)).mean()
+            self.alpha_optimizer.zero_grad()
+            alpha_loss.backward()
+            self.alpha_optimizer.step()
+            tabular.record("alpha", torch.exp(self.log_alpha.detach()).item())
+            alpha = self.log_alpha.detach().exp().item()
+        else:
+            alpha = self.alpha
+        min_q_new_actions = torch.min(self.qf1(torch.Tensor(obs), torch.Tensor(new_actions)), 
+                            self.qf2(torch.Tensor(obs), torch.Tensor(new_actions)))
+        policy_objective = ((alpha * log_pi) - min_q_new_actions.flatten()).mean()
+        tabular.record("policy_loss", policy_objective.item())
+
+        '''
+        QF Loss
```

Comment:
these block comments are a sure sign that this function should be split up into private subroutines.